File: cpp/example_code/cloudwatch/cwl_describe_subscription_filters.cpp
/*
   Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.

   This file is licensed under the Apache License, Version 2.0 (the "License").
   You may not use this file except in compliance with the License. A copy of
   the License is located at

    http://aws.amazon.com/apache2.0/

   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
   CONDITIONS OF ANY KIND, either express or implied. See the License for the
   specific language governing permissions and limitations under the License.
*/
#include <aws/core/Aws.h>

#include <aws/core/utils/Outcome.h>

#include <aws/logs/CloudWatchLogsClient.h>
#include <aws/logs/model/DescribeSubscriptionFiltersRequest.h>
#include <aws/logs/model/DescribeSubscriptionFiltersResult.h>

#include <iostream>
#include <iomanip>

/**
 * Lists cloudwatch subscription filters associated with a log group
 */
int main(int argc, char** argv)
{
    if (argc != 2)
    {
        std::cout << "Usage: cw_describe_subscription_filters <log_group_name>" << std::endl;
        return 1;
    }

    Aws::String logGroupName(argv[1]);

    Aws::SDKOptions options;
    Aws::InitAPI(options);

    Aws::CloudWatchLogs::CloudWatchLogsClient cwl_client;

    Aws::CloudWatchLogs::Model::DescribeSubscriptionFiltersRequest describeSubscriptionFiltersRequest;
    describeSubscriptionFiltersRequest.SetLogGroupName(logGroupName);
    describeSubscriptionFiltersRequest.SetLimit(1);

    bool done = false;
    bool header = false;
    while(!done)
    {
        auto describeSubscriptionFiltersOutcome = cwl_client.DescribeSubscriptionFilters(describeSubscriptionFiltersRequest);
        if(!describeSubscriptionFiltersOutcome.IsSuccess())
        {
            std::cout << "Failed to describe cloudwatch subscription filters for log group " << logGroupName << ": " << describeSubscriptionFiltersOutcome.GetError().GetMessage() << std::endl;
            break;
        }

        if(!header)
        {
            std::cout << std::left << std::setw(32) << "Name" 
                                   << std::setw(64) << "FilterPattern"
                                   << std::setw(64) << "DestinationArn" << std::endl;
            header = true;
        }

        const auto& filters = describeSubscriptionFiltersOutcome.GetResult().GetSubscriptionFilters();
        for (const auto& filter : filters)
        {
            std::cout << std::left << std::setw(32) << filter.GetFilterName() 
                                   << std::setw(64) << filter.GetFilterPattern()
                                   << std::setw(64) << filter.GetDestinationArn() << std::endl;
        }

        const auto& nextToken = describeSubscriptionFiltersOutcome.GetResult().GetNextToken();
        describeSubscriptionFiltersRequest.SetNextToken(nextToken);
        done = nextToken.empty();
    }

    Aws::ShutdownAPI(options);

    return 0;
}

File: javascript/example_code/cloudwatch/cwl_describesubscriptionfilters.js
/**
 * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * This file is licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License. A copy of
 * the License is located at
 *
 * http://aws.amazon.com/apache2.0/
 *
 * This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
*/

//snippet-sourcedescription:[cwl_describesubscriptionfilters.js demonstrates how to list the subscription filters for an Amazon CloudWatch Logs group.]
//snippet-service:[cs_logs]
//snippet-keyword:[JavaScript]
//snippet-sourcesyntax:[javascript]
//snippet-keyword:[Code Sample]
//snippet-keyword:[Amazon CloudWatch Logs]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[2018-06-02]
//snippet-sourceauthor:[AWS-JSDG]

// ABOUT THIS NODE.JS SAMPLE: This sample is part of the SDK for JavaScript Developer Guide topic at
// https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/cloudwatch-examples-subscriptions.html 

// snippet-start:[cwLogs.JavaScript.cwl.describeSubscriptionFilters]
// Load the AWS SDK for Node.js
var AWS = require('aws-sdk');
// Set the region 
AWS.config.update({region: 'REGION'});

// Create the CloudWatchLogs service object
var cwl = new AWS.CloudWatchLogs({apiVersion: '2014-03-28'});

var params = {
  logGroupName: 'GROUP_NAME',
  limit: 5
};

cwl.describeSubscriptionFilters(params, function(err, data) {
  if (err) {
    console.log("Error", err);
  } else {
    console.log("Success", data.subscriptionFilters);
  }
});
// snippet-end:[cwLogs.JavaScript.cwl.describeSubscriptionFilters]
